src/audio/copier/copier_dai.c

```diff
@@ -300,3 +300,156 @@ int copier_dai_prepare(struct comp_dev *dev, struct copier_data *cd)
+
+	/* We have to iterate over frames here. However, tracking frames requires using
+	 * of expensive division operations (e.g., inside audio_stream_frames_without_wrap()).
+	 * So let's track samples instead. Since we only copy one channel, src_stream_sample_count
+	 * is NOT number of samples we need to copy but total samples for all channels. We just
+	 * track them to know when to stop.
+	 */
+	int src_stream_sample_count = frame_count * audio_stream_get_channels(src);
+
+	while (src_stream_sample_count) {
+		int src_samples_without_wrap;
+		int16_t *r_end_ptr, *r_ptr_before_loop;
+
+		r_ptr = audio_stream_wrap(src, r_ptr);
+		w_ptr = audio_stream_wrap(dst, w_ptr);
+
+		src_samples_without_wrap = audio_stream_samples_without_wrap_s16(src, r_ptr);
+		r_end_ptr = src_stream_sample_count < src_samples_without_wrap ?
+			r_ptr + src_stream_sample_count : (int16_t *)audio_stream_get_end_addr(src);
+
+		r_ptr_before_loop = r_ptr;
+
+		do {
+			*w_ptr = *r_ptr;
+			r_ptr += audio_stream_get_channels(src);
+			w_ptr += audio_stream_get_channels(dst);
+		} while (r_ptr < r_end_ptr && w_ptr < (int16_t *)audio_stream_get_end_addr(dst));
+
+		src_stream_sample_count -= r_ptr - r_ptr_before_loop;
+	}
+
+	return 0;
+}
+
+static int copy_single_channel_c32(const struct audio_stream __sparse_cache *src,
+				   unsigned int src_channel,
+				   struct audio_stream __sparse_cache *dst,
+				   unsigned int dst_channel, unsigned int frame_count)
+{
+	int32_t *r_ptr = (int32_t *)audio_stream_get_rptr(src) + src_channel;
+	int32_t *w_ptr = (int32_t *)audio_stream_get_wptr(dst) + dst_channel;
+
+	/* We have to iterate over frames here. However, tracking frames requires using
+	 * of expensive division operations (e.g., inside audio_stream_frames_without_wrap()).
+	 * So let's track samples instead. Since we only copy one channel, src_stream_sample_count
+	 * is NOT number of samples we need to copy but total samples for all channels. We just
+	 * track them to know when to stop.
+	 */
+	int src_stream_sample_count = frame_count * audio_stream_get_channels(src);
+
+	while (src_stream_sample_count) {
+		int src_samples_without_wrap;
+		int32_t *r_end_ptr, *r_ptr_before_loop;
+
+		r_ptr = audio_stream_wrap(src, r_ptr);
+		w_ptr = audio_stream_wrap(dst, w_ptr);
+
+		src_samples_without_wrap = audio_stream_samples_without_wrap_s32(src, r_ptr);
+		r_end_ptr = src_stream_sample_count < src_samples_without_wrap ?
+			r_ptr + src_stream_sample_count : (int32_t *)audio_stream_get_end_addr(src);
+
+		r_ptr_before_loop = r_ptr;
+
+		do {
+			*w_ptr = *r_ptr;
+			r_ptr += audio_stream_get_channels(src);
+			w_ptr += audio_stream_get_channels(dst);
+		} while (r_ptr < r_end_ptr && w_ptr < (int32_t *)audio_stream_get_end_addr(dst));
+
+		src_stream_sample_count -= r_ptr - r_ptr_before_loop;
+	}
+
+	return 0;
+}
+
+int copier_dai_params(struct sof_ipc_stream_params *params,
+		      struct comp_dev *dev, struct copier_data *cd, int dai_index)
```

Comment:
don't all these exported copier functions have `dev` as their first parameter probably followed by `cd`? Would be good to follow some such convention

Reply:
yes, I like put cd first, dev as second, since dev is only for log output, let me change.